Firebase/CoreDiagnostics/FIRCDLibrary/FIRCoreDiagnostics.m

```diff
@@ -0,0 +1,688 @@
+
+// Reconcile this with kFIRLoggerInAppMessaging defined in FIRCore+InAppMessaging.m
+// after fiam sdk is released
+static NSString *const kFIRServiceIAM = @"InAppMessaging";
+
+/**
+ * The file name to keep the unique install string.
+ */
+static NSString *const kUniqueInstallFileName = @"FIREBASE_UNIQUE_INSTALL";
+
+/**
+ * @note This should implement the GDTEventDataObject protocol, but can't because of weak-linking.
+ */
+@interface FIRCoreDiagnosticsLog : NSObject
+
+/** The config that will be converted to proto bytes. */
+@property(nonatomic) logs_proto_mobilesdk_ios_ICoreConfiguration config;
+
+@end
+
+@implementation FIRCoreDiagnosticsLog
+
+- (instancetype)initWithConfig:(logs_proto_mobilesdk_ios_ICoreConfiguration)config {
+  self = [super init];
+  if (self) {
+    _config = config;
+  }
+  return self;
+}
+
+// Provided and required by the GDTEventDataObject protocol.
+- (NSData *)transportBytes {
+  pb_ostream_t sizestream = PB_OSTREAM_SIZING;
+
+  // Encode 1 time to determine the size.
+  if (!pb_encode(&sizestream, logs_proto_mobilesdk_ios_ICoreConfiguration_fields, &_config)) {
+    NSCAssert(NO, @"Error in nanopb encoding for size: %s", PB_GET_ERROR(&sizestream));
+  }
+
+  // Encode a 2nd time to actually get the bytes from it.
+  size_t bufferSize = sizestream.bytes_written;
+  CFMutableDataRef dataRef = CFDataCreateMutable(CFAllocatorGetDefault(), bufferSize);
+  pb_ostream_t ostream = pb_ostream_from_buffer((void *)CFDataGetBytePtr(dataRef), bufferSize);
+  if (!pb_encode(&ostream, logs_proto_mobilesdk_ios_ICoreConfiguration_fields, &_config)) {
+    NSCAssert(NO, @"Error in nanopb encoding for bytes: %s", PB_GET_ERROR(&ostream));
+  }
+  CFDataSetLength(dataRef, ostream.bytes_written);
+
+  return CFBridgingRelease(dataRef);
+}
+
+- (void)dealloc {
+  pb_release(logs_proto_mobilesdk_ios_ICoreConfiguration_fields, &_config);
+}
+
+@end
+
+/** This class produces a protobuf containing diagnostics and usage data to be logged. */
+@interface FIRCoreDiagnostics : NSObject <FIRCoreDiagnosticsInterop>
+
+/** The queue on which all diagnostics collection will occur. */
+@property(nonnull, nonatomic) dispatch_queue_t diagnosticsQueue;
+
+/** The transport object used to send data. */
+@property(nonnull, nonatomic) GDTTransport *transport;
+
+@end
+
+@implementation FIRCoreDiagnostics
+
++ (void)load {
+  FIRCoreDiagnosticsImplementation = self;
+}
+
++ (NSString *)libraryVersion {
+  return @STR(FIRCoreDiagnostics_VERSION);
+}
+
++ (instancetype)sharedInstance {
+  static FIRCoreDiagnostics *sharedInstance;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    sharedInstance = [[FIRCoreDiagnostics alloc] init];
+  });
+  return sharedInstance;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _diagnosticsQueue =
+        dispatch_queue_create("com.google.FIRCoreDiagnostics", DISPATCH_QUEUE_SERIAL);
+    _transport = [[GDTTransport alloc] initWithMappingID:@"137"
```

Comment:
I wonder, why @"137"?

Reply:
It's the log source enum in clientanalytics.proto. We use ints instead of enums to save binary size.